### PR TITLE
Better database connection handling for connectors

### PR DESCRIPTION
**Pull request by jdpgrailsdev**
## What
* Ensures that connections to the database are being properly closed 

## How
* Ensure inversion of control for all connector code that interacts with a `Database` or `DataSource` object.

## Recommended reading order
1. `AbstractDatabase.java`
2. `DataSourceFactory.java`
3.  All other classes

## 🚨 User Impact 🚨
There should be no user impact.

## Tests

<details><summary><strong>Unit</strong></summary>

* All platform unit tests pass
* All connector unit tests pass

</details>

### CODE REVIEW

File: airbyte-db/lib/src/main/java/io/airbyte/db/instance/BaseDatabaseInstance.java
Comment (davinchia) [inline airbyte-db/lib/src/main/java/io/airbyte/db/instance/BaseDatabaseInstance.java:113]: nice!

File: airbyte-db/lib/src/main/java/io/airbyte/db/AbstractDatabase.java
Comment (davinchia) [inline airbyte-db/lib/src/main/java/io/airbyte/db/AbstractDatabase.java:9]: why did you remove this from here?

Comment (jdpgrailsdev) [reply airbyte-db/lib/src/main/java/io/airbyte/db/AbstractDatabase.java:9]: The database class is no longer what is closeable.  It is merely now a wrapper around the instantiated `DataSource`, which is what we actually want to close.  By allowing the `AbstractDatabse` to close the data source, we might accidentally close a shared data source between multiple instances.  Therefore, it is the data source that should be directly closed, removing the need for this class to implement closeable (it has nothing to close now that inversion of control has been enforced).

**Review by davinchia — COMMENTED** [review]
This generally looks sane to me. I haven't looked at destination code for half a year now, so I want to defer to @tuliren for the final approval. There are parts in the main database classes I'm not sure are closing the connector (though this might not be related to your change).

Since we are modifying the connectors, we probably want to bump their version and release a new connector version - exciting! Will also defer to Liren on that.

File: airbyte-integrations/connectors/destination-databricks/src/main/java/io/airbyte/integrations/destination/databricks/DatabricksDestination.java
Comment (jdpgrailsdev) [reply airbyte-integrations/connectors/destination-databricks/src/main/java/io/airbyte/integrations/destination/databricks/DatabricksDestination.java:66]: Good question.  Doesn't seem like it.  I am not too familiar with this code.  Ideally, we would want whatever runs the source/destination to close the DataSource once it completes.  For what it's worth, I don't believe the original `JdbcDatabase` was closed either.  I think that this would need to be redesigned in order to do so.

File: airbyte-db/lib/src/main/java/io/airbyte/db/AbstractDatabase.java
Comment (tuliren) [reply airbyte-db/lib/src/main/java/io/airbyte/db/AbstractDatabase.java:9]: This explanation is helpful. Maybe copy some of that as the Javadoc of this class? Something like:

```java
/**
 * The Database class is a wrapper around the instantiated DataSource. The DataSource is closeable, but
 * not the Database class. This is because we don't want to accidentally close a shared data source between
 * multiple instances by closing the Database class.
 */
```

File: airbyte-db/lib/src/main/java/io/airbyte/db/factory/DataSourceFactory.java
Comment (tuliren) [inline airbyte-db/lib/src/main/java/io/airbyte/db/factory/DataSourceFactory.java:161]: Probably safer to use `AutoCloseable` because 1) `Closeable` extends from `AutoCloseable`, and 2) the common `BasicDataSource` we use in lots of places is an `AutoCloseable` but not a `Closeable`.

```suggestion
      if (dataSource instanceof AutoCloseable closeable) {
```

File: airbyte-integrations/connectors/destination-databricks/src/main/java/io/airbyte/integrations/destination/databricks/DatabricksDestination.java
Comment (tuliren) [reply airbyte-integrations/connectors/destination-databricks/src/main/java/io/airbyte/integrations/destination/databricks/DatabricksDestination.java:66]: I think the Databricks database is closed somewhere. Previously it was not properly closed, and the acceptance tests would just hang there.

File: airbyte-integrations/connectors/destination-databricks/src/test-integration/java/io/airbyte/integrations/destination/databricks/DatabricksDestinationAcceptanceTest.java
Comment (tuliren) [inline airbyte-integrations/connectors/destination-databricks/src/test-integration/java/io/airbyte/integrations/destination/databricks/DatabricksDestinationAcceptanceTest.java:147]: The helper method here did not seem too bad, since it was referenced twice. Also it made the `try` statement shorter.

**Review by tuliren — APPROVED** [review]

**Comment by tuliren** [comment]
> There are parts in the main database classes I'm not sure are closing the connector (though this might not be related to your change).

I think at this point, all database connections should be closed.

> Since we are modifying the connectors, we probably want to bump their version and release a new connector version

@jdpgrailsdev, since there is no new feature or bug fix, we don't need to bump the connector versions. We do need to make sure all the connectors are functioning normally. Can you trigger integration tests for all affected connector on this PR?

File: airbyte-integrations/connectors/destination-databricks/src/main/java/io/airbyte/integrations/destination/databricks/DatabricksDestination.java
Comment (tuliren) [reply airbyte-integrations/connectors/destination-databricks/src/main/java/io/airbyte/integrations/destination/databricks/DatabricksDestination.java:66]: Confirmed that this is closed. The `getDatabase` method is used in `CopyDestination` in a `try` statement:

https://github.com/airbytehq/airbyte/blob/v0.38.2-alpha/airbyte-integrations/connectors/destination-jdbc/src/main/java/io/airbyte/integrations/destination/jdbc/copy/CopyDestination.java#L57

```java
    try (final JdbcDatabase database = getDatabase(config)) {
      final var nameTransformer = getNameTransformer();
      final var outputSchema = nameTransformer.convertStreamName(config.get(schemaFieldName).asText());
      AbstractJdbcDestination.attemptSQLCreateAndDropTableOperations(outputSchema, database, nameTransformer, getSqlOperations());

      return new AirbyteConnectionStatus().withStatus(AirbyteConnectionStatus.Status.SUCCEEDED);
    }
```

**Comment by jdpgrailsdev** [comment]
> > There are parts in the main database classes I'm not sure are closing the connector (though this might not be related to your change).
> 
> I think at this point, all database connections should be closed.
>

Yes.  This PR does not change that.  What is being questioned is if the data source from which those connections come from are closed.  The existing code did not always close the data source when done with it, so I believe that the proposed changes will not make that any worse.

> > Since we are modifying the connectors, we probably want to bump their version and release a new connector version
> 
> @jdpgrailsdev, since there is no new feature or bug fix, we don't need to bump the connector versions. We do need to make sure all the connectors are functioning normally. Can you trigger integration tests for all affected connector on this PR?
> 

@tuliren Sure.  Could you please point me to the documentation regarding how to trigger the integration tests if they are not already part of the checks the run as part of the PR build?

**Comment by tuliren** [comment]
> What is being questioned is if the data source from which those connections come from are closed. The existing code did not always close the data source when done with it, so I believe that the proposed changes will not make that any worse.

I think that all data sources are currently closed (before this PR). For example, see my comment [here](https://github.com/airbytehq/airbyte/pull/12743#discussion_r870752993) about the `getDatabase` in the Databricks connector.

> @tuliren Sure. Could you please point me to the documentation regarding how to trigger the integration tests if they are not already part of the checks the run as part of the PR build?

Here is the doc: https://docs.airbyte.com/connector-development/testing-connectors/#3-requesting-github-pr-integration-test-runs

Here are the command for the affected connectors:

```
/test connector=connectors/destination-clickhouse
/test connector=connectors/destination-csv
/test connector=connectors/destination-databricks
/test connector=connectors/destination-mariadb-columnstore
/test connector=connectors/destination-mysql
/test connector=connectors/destination-oracle
/test connector=connectors/destination-redshift
/test connector=connectors/destination-snowflake
/test connector=connectors/source-clickhouse
/test connector=connectors/source-cockroachdb
/test connector=connectors/source-db2
/test connector=connectors/source-mongodb
/test connector=connectors/source-mssql
/test connector=connectors/source-mysql
/test connector=connectors/source-oracle
/test connector=connectors/source-snowflake
```

On a second thought, given the similarity of the changes, we can just test a few destinations and a few sources. These are the important ones:

```
/test connector=connectors/destination-databricks
/test connector=connectors/destination-mysql
/test connector=connectors/destination-oracle
/test connector=connectors/destination-redshift
/test connector=connectors/destination-snowflake

/test connector=connectors/source-mongodb
/test connector=connectors/source-mssql
/test connector=connectors/source-mysql
/test connector=connectors/source-oracle
/test connector=connectors/source-snowflake
```

File: airbyte-integrations/connectors/destination-databricks/src/main/java/io/airbyte/integrations/destination/databricks/DatabricksDestination.java
Comment (jdpgrailsdev) [reply airbyte-integrations/connectors/destination-databricks/src/main/java/io/airbyte/integrations/destination/databricks/DatabricksDestination.java:66]: > 

@tuliren That is still the case in the refactored code.  Instead of using the try-with-resources, it now uses a try/finally approach, as the Java `DataSource` interface unfortunately does not implement `AutoCloseable` or `Closeable`.  The areas that are being questioned are ones like this:

https://github.com/airbytehq/airbyte/blob/master/airbyte-integrations/connectors/destination-databricks/src/main/java/io/airbyte/integrations/destination/databricks/DatabricksDestination.java#L41

The existing code gets a `Database` instance, but it is not clear that it is ever closed.  Perhaps it is eventually closed somewhere in the calling code?  I believe that this is the only place that @davinchia and I had questions about with regards to whether or not the changes in this PR match the existing behavior.

**Comment by jdpgrailsdev** [comment]
@tuliren I have addressed all of the comments from your review.  It looks like there may be some failing tests that have leaked in from master after I rebased to resolve some conflicts.  However, I don't have the context of those tests to be able to definitively say if they are related to my changes.  Let me know if there is something I need to look into.  Thanks.

File: airbyte-integrations/connectors/destination-databricks/src/main/java/io/airbyte/integrations/destination/databricks/DatabricksDestination.java
Comment (jdpgrailsdev) [reply airbyte-integrations/connectors/destination-databricks/src/main/java/io/airbyte/integrations/destination/databricks/DatabricksDestination.java:69]: @tuliren and I paired on this and we have resolved the issue with the DataSource not being closed.

File: airbyte-integrations/connectors/destination-databricks/src/test-integration/java/io/airbyte/integrations/destination/databricks/DatabricksDestinationAcceptanceTest.java
Comment (jdpgrailsdev) [reply airbyte-integrations/connectors/destination-databricks/src/test-integration/java/io/airbyte/integrations/destination/databricks/DatabricksDestinationAcceptanceTest.java:147]: This has now been resolved.

**Pull request merged**
